### PR TITLE
feat: interactive legends

### PR DESCRIPTION
**Pull request by arvind**
This PR supersedes #4732 and makes legends interactive when there exists a selection that has been projected over the legend's domain. If the legend only partially matches the selection's projection (e.g., a color legend for `Origin` when the selection is projected over `["Origin", "Cylinders"]`) the legend is still interactive and only partially populates the selection predicate (i.e., clicking "Europe" will select all "Europe" cars without any "Cylinders" specified.)

Kapture 2019-08-15 at 17 00 59

/cc @djbarnwal

### CODE REVIEW

**Comment by domoritz** [comment]
🎉 I'm very excited about this

**Comment by arvind** [comment]
Alright, I'm nearing the end with this feature PR. Before I finalize the design via unit tests, I'd like to get some eyes and hands on how interactive legends work and confirm that the current design is desirable. 

To that end, here are some interesting specifications to try. I'm purposely not including any GIFs so as not to prime you. 

<details><summary>Single selection projected over the legend domain</summary>

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v3.json",
  "description": "A scatterplot showing horsepower and miles per gallons.",
  "data": {"url": "data/cars.json"},
  "mark": "circle",
  "selection": {
    "foo": {"type": "single", "fields": ["Origin"]}
  },
  "encoding": {
    "x": {"field": "Horsepower", "type": "quantitative"},
    "y": {"field": "Miles_per_Gallon", "type": "quantitative"},
    "color": {
      "condition": {"selection": "foo", "field": "Origin", "type": "nominal"},
      "value": "grey"
    }
  }
}
```

</details>

<details><summary>Multi selection projected over the legend domain</summary>

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v3.json",
  "description": "A scatterplot showing horsepower and miles per gallons.",
  "data": {"url": "data/cars.json"},
  "mark": "circle",
  "selection": {
    "foo": {"type": "single", "fields": ["Origin"]}
  },
  "encoding": {
    "x": {"field": "Horsepower", "type": "quantitative"},
    "y": {"field": "Miles_per_Gallon", "type": "quantitative"},
    "color": {
      "condition": {"selection": "foo", "field": "Origin", "type": "nominal"},
      "value": "grey"
    }
  }
}
```

</details>

<details><summary>Multi selection projected over several fields, only one of which includes the legend domain</summary>

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v3.json",
  "description": "A scatterplot showing horsepower and miles per gallons.",
  "data": {"url": "data/cars.json"},
  "mark": "circle",
  "selection": {
    "foo": {"type": "multi", "fields": ["Origin", "Cylinders"]}
  },
  "encoding": {
    "x": {"field": "Horsepower", "type": "quantitative"},
    "y": {"field": "Miles_per_Gallon", "type": "quantitative"},
    "color": {
      "condition": {"selection": "foo", "field": "Origin", "type": "nominal"},
      "value": "grey"
    }
  }
}
```

</details>

<details><summary>Multi selection projected over several fields, where each field has a corresponding legend</summary>

```json
{
  "$schema": "htts://vega.github.io/schema/vega-lite/v3.json",
  "description": "A scatterplot showing horsepower and miles per gallons.",
  "data": {"url": "data/cars.json"},
  "mark": "circle",
  "selection": {
    "foo": {"type": "multi", "fields": ["Origin", "Cylinders"]}
  },
  "encoding": {
    "x": {"field": "Horsepower", "type": "quantitative"},
    "y": {"field": "Miles_per_Gallon", "type": "quantitative"},
    "color": {
      "condition": {"selection": "foo", "field": "Origin", "type": "nominal"},
      "value": "grey"
    },
    "size": {
      "condition": {"selection": "foo", "field": "Cylinders", "type": "nominal"},
      "value": 0
    }
  }
}
```

</details>

<details><summary>Multiple selections each that map to one legend</summary>

```json
{
  "$schema": "htts://vega.github.io/schema/vega-lite/v3.json",
  "description": "A scatterplot showing horsepower and miles per gallons.",
  "data": {"url": "data/cars.json"},
  "mark": "circle",
  "selection": {
    "foo": {"type": "multi", "fields": ["Origin"]},
    "bar": {"type": "multi", "fields": ["Cylinders"]}
  },
  "encoding": {
    "x": {"field": "Horsepower", "type": "quantitative"},
    "y": {"field": "Miles_per_Gallon", "type": "quantitative"},
    "color": {
      "condition": {"selection": "foo", "field": "Origin", "type": "nominal"},
      "value": "grey"
    },
    "size": {
      "condition": {"selection": "bar", "field": "Cylinders", "type": "nominal"},
      "value": 0
    }
  }
}
```

</details>

<details><summary>Interactive Legends when unioning a selection across multiple views</summary>

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v3.json",
  "repeat": {
    "row": ["Horsepower", "Acceleration", "Miles_per_Gallon"],
    "column": ["Miles_per_Gallon", "Acceleration", "Horsepower"]
  },
  "spec": {
    "data": {"url": "data/cars.json"},
    "mark": "circle",
    "selection": {
      "brush": {
        "type": "multi", "fields": ["Origin"],
        "resolve": "union"
      }
    },
    "encoding": {
      "x": {"field": {"repeat": "column"}, "type": "quantitative"},
      "y": {
        "field": {"repeat": "row"},
        "type": "quantitative",
        "axis": {"minExtent": 30}
      },
      "color": {
        "condition": {
          "selection": "brush",
          "field": "Origin",
          "type": "nominal"
        },
        "value": "grey"
      }
    }
  }
}

```

</details>

Let me know what you think. Besides bugs, do interactive legends work as you expect them to or are there cases where their behavior is unexpected?

/cc @djbarnwal, @domoritz, @kanitw, @jheer

**Comment by domoritz** [comment]
* This is awesome!
* I think the second spec should use `multi`, not `single` as the title implies. 
* In the multi-selection, I first wanted to use CMD since I use that on my mac to toggle selections but it didn't work. This obviously isn't something unique to this PR.
* I expected shift+click to also de-select a previously selected item but it didn't work for me. Is this supposed to work? It works in the chart but not the legend. 
* The way multiple legends over a single projection work is really nice!

**Comment by domoritz** [comment]
Are we changing the design based on @jheer's comment in https://github.com/vega/vega-lite/issues/1657#issuecomment-530436526?

**Comment by domoritz** [comment]
Can you rebase and squash?

File: examples/compiled/area.vg.json [hunk 1]
```diff
@@ -114,5 +114,6 @@
       "tickCount": {"signal": "ceil(height/40)"},
       "zindex": 0
     }
-  ]
+  ],
+  "config": {"legend": {"unselectedOpacity": 0.35}}
```
Comment (domoritz) [inline examples/compiled/area.vg.json]: Hmm, this shouldn't be here.

Comment (domoritz) [reply examples/compiled/area.vg.json]: Removed

File: src/compile/selection/transforms/legends.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+    for (const proj of selCmpt.project.items) {
+      if (!proj.hasLegend) continue;
+      const prefix = `${proj.field}_legend`;
+      const sgName = `${selName}_${prefix}`;
+      const hasSignal = signals.filter(s => s.name === sgName);
+
+      if (!hasSignal.length) {
+        const events = stream.merge
+          .map(markName(`${prefix}_symbols`))
```
Comment (domoritz) [inline src/compile/selection/transforms/legends.ts]: The prefix can be undefined for legends that show counts.

File: examples/compiled/bar_layered_transparent.vg.json [hunk 3]
```diff
@@ -117,7 +117,13 @@
       "gradientLength": {"signal": "clamp(height, 64, 200)"},
       "symbolType": "square",
       "title": "gender",
-      "encode": {"symbols": {"update": {"opacity": {"value": 0.7}}}}
+      "encode": {
+        "symbols": {
+          "name": "gender_legend_symbols",
```
Comment (domoritz) [inline examples/compiled/bar_layered_transparent.vg.json]: This chart isn't interactive so it shouldn't have a name for the legend, right?

**Comment by domoritz** [comment]
This spec leads to an error "Converting circular structure to JSON"

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
  "data": {"url": "data/cars.json"},
  "selection": {
    "sel": {
      "type": "multi", "fields": ["Origin"], "bind": "legend"
    }
  },
  "mark": "point",
  "encoding": {
    "x": {"field": "Horsepower", "type": "quantitative"},
    "y": {"field": "Miles_per_Gallon", "type": "quantitative"},
    "shape": {"field": "Origin", "type": "nominal"},
    "color": {
      "condition": {"selection": "sel", "field": "Origin", "type": "nominal"},
      "value": "gray"
    }
  }
}
```

**Comment by domoritz** [comment]
In this spec below, I would expect the shape to depend on the selection. Instead, Vega-Lite filters the data by the selection. I think the actual behavior makes a lot more sense but it's not what the spec says, no?

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
  "data": {"url": "data/cars.json"},
  "selection": {
    "sel": {
      "type": "multi", "fields": ["Origin"], "bind": "legend"
    }
  },
  "mark": "point",
  "encoding": {
    "x": {"field": "Horsepower", "type": "quantitative"},
    "y": {"field": "Miles_per_Gallon", "type": "quantitative"},
    "shape": {
      "condition": {"selection": "sel", "field": "Origin", "type": "nominal"},
      "value": "rectangle"
    }
  }
}

```

**Comment by domoritz** [comment]
https://github.com/vega/vega-lite/pull/5305#issuecomment-554862968 seems to be happening in legend assemble.

Screen Shot 2019-11-17 at 21 58 46

Works in current Vega-Lite: https://vega.github.io/editor/#/url/vega-lite/N4KABGBEAkDODGALApgWwIaQFxUQFzwAdYsB6UgN2QHN0A6agSz0QFcAjOxge1IRQyUa6ALQAbZskoAWOgCtY3AHaQANOCgATdHkw5gkVgCcx2LTvSl46I7HmKVAX3URIsZGOTw8PFfo2u7qb+EKFQeACehMhmkKisYj5qUABmjB6asGYA2pAA8kaMTCoAuqpQ7IxKmrGe1MjVkAFgjhrOGnE2ANaxhNxVeGodDfDcmlXUZqBhkAAeU5BpGbEAEty2yH0A7shGyZCR0bEAjqzoSj66PlSQ7TMRC0tiNTiQALKMnrAA+tFG3wBxdBiMTKfaHGKvU7nS46Rg3O6hNyIdBHfSLdLPWIFIpVcFRSFQJTcVBVYG3FxI0agvYhMJQUbVZi+BZBLw+MGvIL7J4vKA44r4tFEklk0yImYUYGsQmQahGdAPZqtCCtRxAA

**Comment by domoritz** [comment]
Fixed in https://github.com/vega/vega-lite/pull/5305/commits/17bb58d413ff262d3e56076d3af8e016679bf2bf

File: src/compile/selection/transforms/legends.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+export default legendBindings;
+
+export function parseInteractiveLegend(
+  model: UnitModel,
+  channel: NonPositionScaleChannel,
+  legendCmpt: LegendComponent
+) {
+  const field = model.fieldDef(channel).field;
+  forEachSelection(model, selCmpt => {
+    const proj = selCmpt.project.hasField[field] || selCmpt.project.hasChannel[channel];
```
Comment (domoritz) [inline src/compile/selection/transforms/legends.ts]:
```suggestion
    const proj = selCmpt.project.hasField[field] ?? selCmpt.project.hasChannel[channel];
```

File: src/compile/selection/transforms/legends.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    if (proj && legendBindings.has(selCmpt)) {
+      const legendSelections = legendCmpt.get('selections') || [];
```
Comment (domoritz) [inline src/compile/selection/transforms/legends.ts]:
```suggestion
      const legendSelections = legendCmpt.get('selections') ?? [];
```

File: src/compile/legend/encode.ts [hunk 6]
```diff
@@ -94,8 +96,9 @@ export function symbols(
   }
 
   if (channel !== OPACITY) {
-    if (opacity) {
-      // only apply opacity if it is neither zero or undefined
+    if (condition) {
+      out.opacity = [{test: condition, value: opacity || 1}, {value: config.legend.unselectedOpacity}];
```
Comment (domoritz) [inline src/compile/legend/encode.ts]:
```suggestion
      out.opacity = [{test: condition, value: opacity ?? 1}, {value: config.legend.unselectedOpacity}];
```

File: src/compile/legend/parse.ts [hunk 7]
```diff
@@ -102,15 +104,20 @@ export function parseLegendForChannel(model: UnitModel, channel: NonPositionScal
     }
   }
 
-  const legendEncoding = legend.encoding ?? {};
-  const legendEncode = (['labels', 'legend', 'title', 'symbols', 'gradient'] as const).reduce(
+  const legendEncoding = legend.encoding || {};
```
Comment (domoritz) [inline src/compile/legend/parse.ts]:
```suggestion
  const legendEncoding = legend.encoding ?? {};
```

File: src/compile/selection/transforms/legends.ts [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+  parse: (model, selCmpt, selDef, origDef) => {
+    // Binding a selection to a legend disables default direct manipulation interaction.
+    // A user can choose to re-enable it by explicitly specifying triggering input events.
+    if (!origDef.on) delete selCmpt.events;
+    if (!origDef.clear) delete selCmpt.clear;
+
+    if (origDef.on || origDef.clear) {
+      const legendFilter = 'event.item && indexof(event.item.mark.role, "legend") < 0';
+      for (const evt of selCmpt.events) {
+        evt.filter = array(evt.filter || []);
```
Comment (domoritz) [inline src/compile/selection/transforms/legends.ts]:
```suggestion
        evt.filter = array(evt.filter ?? []);
```

File: src/compile/selection/transforms/legends.ts (hunk 2 above)
Comment (arvind) [reply src/compile/selection/transforms/legends.ts]: Selections are not supported for count/aggregate fields, so with https://github.com/vega/vega-lite/pull/5305/commits/d05d65126d851e97458736c76600ea86d07db605 we warn and skip these types of definitions. Moreover, as of https://github.com/vega/vega-lite/pull/5305/commits/e6917e25c6ecb4c292e497bb5ef9bfd3db1a6758, we only output legend names/interactive flags if a corresponding selection is found. Together, these two commits should address this point.

**Pull request merged**